modules/fedimint-ln-server/src/lib.rs

```diff
@@ -885,11 +919,60 @@ impl ServerModule for Lightning {
 }
 
 impl Lightning {
-    pub fn new(cfg: LightningConfig, task_group: &mut TaskGroup) -> anyhow::Result<Self> {
+    pub async fn new(
+        cfg: LightningConfig,
+        db: Database,
+        task_group: &mut TaskGroup,
+    ) -> anyhow::Result<Self> {
         let btc_rpc = create_bitcoind(&cfg.local.bitcoin_rpc, task_group.make_handle())?;
+        task_group
+            .spawn("compact audit keys", |handle| async move {
+                Self::compact_audit_keys_task(db, &handle).await;
+            })
+            .await;
         Ok(Lightning { cfg, btc_rpc })
     }
 
+    async fn compact_audit_keys_task(db: Database, tg_handle: &TaskHandle) {
+        let interval = Duration::from_secs(60 * 5);
+        while !tg_handle.is_shutting_down() {
+            let mut dbtx = db.begin_transaction().await;
+            Self::compact_audit_keys(&mut dbtx).await;
+            if dbtx.commit_tx_result().await.is_err() {
+                warn!(
+                    "conflict while compacting audit keys, retrying in {} seconds",
+                    interval.as_secs()
+                );
+            }
+            sleep(interval).await;
+        }
+    }
+
+    /// Runs compaction for LightningAuditItemKeys
+    ///
+    /// When a contract reaches a terminal state, the associated amount will be
+    /// updated to 0. At this point, the contract no longer needs to be tracked
+    /// for auditing liabilities, so we can safely remove the audit key.
+    async fn compact_audit_keys(dbtx: &mut DatabaseTransaction<'_>) {
+        let _timing /* logs on drop */ = TimeReporter::new("compact_audit_keys");
+        let remove_audit_keys = dbtx
+            .find_by_prefix(&LightningAuditItemKeyPrefix)
+            .await
+            .filter_map(|(key, amount)| async move {
+                if amount.msats == 0u64 {
+                    Some(key)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>()
+            .await;
+
+        for key in remove_audit_keys {
```

Comment:
Could this collection ever get big enough to use a stream and `.for_each(/* remove */)` instead of `collect`?

Reply:
That's a great point and led to an interesting puzzle. The compaction background task has been removed so it's no longer relevant for this PR, but for future reference do you know of a way to reference the dbtx inside of a stream that begins from a dbtx?

I didn't continue investigating, but perhaps it's possible to create an additional `DatabaseTransaction` method that's similar to `find_by_prefix` but also returns the dbtx as part of the stream?

This is the implementation and compiler error.

```rust
        dbtx.find_by_prefix(&LightningAuditItemKeyPrefix)
            .await
            .filter_map(|(key, amount)| async move {
                if amount.msats == 0u64 {
                    Some(key)
                } else {
                    None
                }
            })
            .for_each(|key| async move {
                dbtx.remove_entry(&key);
            })
            .await;
```

```
error[E0507]: cannot move out of `dbtx`, a captured variable in an `FnMut` closure
    --> modules/fedimint-ln-server/src/lib.rs:1007:29
     |
956  |       async fn compact_audit_keys(dbtx: &mut DatabaseTransaction<'_>) {
     |                                   ---- captured outer variable
...
1007 |               .for_each(|key| async move {
     |  _______________________-----_^
     | |                       |
     | |                       captured by this `FnMut` closure
1008 | |                 dbtx.remove_entry(&key);
     | |                 ----
     | |                 |
     | |                 variable moved due to use in generator
     | |                 move occurs because `dbtx` has type `&mut DatabaseTransaction<'_>`, which does not implement the `Copy` trait
1009 | |             })
     | |_____________^ `dbtx` is moved here
```

Perhaps something similar to `.fold` may work if we added a method to `DatabaseTransaction` that returns the results of `find_by_prefix` along with the original dbtx?

```rust
        dbtx.find_by_prefix(&LightningAuditItemKeyPrefix)
            .await
            .filter_map(|(key, amount)| async move {
                if amount.msats == 0u64 {
                    Some(key)
                } else {
                    None
                }
            })
            .fold(dbtx, |dbtx, key| async move {
                dbtx.remove_entry(&key);
                dbtx
            })
            .await;
```

```
error[E0505]: cannot move out of `dbtx` because it is borrowed
    --> modules/fedimint-ln-server/src/lib.rs:1011:19
     |
956  |     async fn compact_audit_keys(dbtx: &mut DatabaseTransaction<'_>) {
     |                                 ---- binding `dbtx` declared here
...
998  |         dbtx.find_by_prefix(&LightningAuditItemKeyPrefix)
     |         ------------------------------------------------- borrow of `*dbtx` occurs here
...
1011 |             .fold(dbtx, |dbtx, key| async move {
     |              ---- ^^^^ move out of `dbtx` occurs here
     |              |
     |              borrow later used by call
```